### PR TITLE
[WIP] Gatsby migration

### PR DESCRIPTION
**Pull request by iAdramelk**
ToDo:

- [x] Install Gatsby.
- [x] Remove Next.js.
- [x] Move `_document.js` content to `html.js` and `SEO` files.
- [x] Rewrite `LocalLink` to use `gatsby-link`.
- [x] Update GA to work with new setup.
- [x] Move sentry from `_app.js` to gatsby plugin.
- [x] Set up `styled-components` SSR in gatsby.
- [x] Move static pages to gatsby (index, features, support, community).
- [x] Replace CDN urls for css and js assets in head with normal imports.
- [x] Docs page
    - [x] Basic render
    - [x] Sidebar
        - [x] Make standard pages work
        - [x] Make `source: false` menu items work (update logic for `normalizedSidebar` generation)
    - [x] Right menu 
        - [x] Headers with ids
        - [x] Custom ids in headers
        - [x] Move `headings` extraction to backend
    - [x] External link icons
    - [x] Prism syntaxes
        - [x] Port custom syntax for `dvc` from blog
        - [x] Convert custom syntax for `usage`
        - [x] Find existing syntax for `vim`
    - [x] Convert dvc commands to links
    - [x] Custom React components for `details` and `abbr`
- [x] Set up redirects.
- [x] Set up API endpoints.

### CODE REVIEW

**Comment by shcheklein** [comment]
@fabiosantoscode could start reviewing this? )

File: src/components/LocalLink/index.js [hunk 1]
```diff
@@ -1,28 +1,20 @@
...
-  as: Component,
-  href,
+  as: SC,
+  href: to,
   ...restProps
 }) {
-  const nextProps = href.match(/^\/doc/)
-    ? { href: PAGE_DOC, as: href }
-    : { href }
+  const Component = SC ? SC.withComponent(GatsbyLink) : GatsbyLink
```
Comment (fabiosantoscode) [inline src/components/LocalLink/index.js]: I'm sorry to say this API is [being deprecated](https://styled-components.com/docs/api#withcomponent)!

The new way to do this is to pass an `as` attribute to the StyledComponent itself as described [here](https://styled-components.com/docs/api#as-polymorphic-prop).

File: src/components/TrySection/index.js [hunk 2]
```diff
@@ -9,14 +9,14 @@ export default function TrySection({ title, buttonText = 'Get Started' }) {
   return (
     <Wrapper>
       <Container>
-        <Glyph src="/static/img/glyph-3.svg" gid={'topleft'} />
+        <Glyph src="/img/glyph-3.svg" gid={'topleft'} />
         <Title>{title}</Title>
         <Buttons>
-          <LocalLink href="/doc/get-started" as={Button} first>
+          <LocalLink href="/doc/get-started" as={Button} first="true">
```
Comment (fabiosantoscode) [inline src/components/TrySection/index.js]: Could you make sure this ends up being rendered as an `<a>`? You probably did, but just making sure for SEO/A11y purposes.

File: src/html.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          key={`body`}
+          id="___gatsby"
+          dangerouslySetInnerHTML={{ __html: props.body }}
+        />
+        {props.postBodyComponents}
+        <script
+          type="text/javascript"
+          src="//downloads.mailchimp.com/js/signup-forms/popup/embed.js"
+          data-dojo-config="usePlainJson: true, isDebug: false"
+        ></script>
```
Comment (fabiosantoscode) [inline src/html.js]: Maybe we could use Helmet to only render this on the page which needs it? (IIRC the home page)

File: src/utils/ga.js [hunk 4]
```diff
@@ -1,24 +1,22 @@
...
 
-export const logException = (description = '', fatal = false) => {
-  if (description) {
-    ReactGA.exception({ description, fatal })
+  if (exDescription) {
+    window.ga('send', {
+      hitType: 'exception',
+      exDescription,
+      exFatal
+    })
```
Comment (fabiosantoscode) [inline src/utils/ga.js]: Aside: We can probably use something like sentry or crashlytics here. If we end up implementing it in another iterative property we can plop it here as well at no additional cost.

File: src/components/Documentation/NewMarkdown/index.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            <span>Prev</span>
+          </LocalLink>
+        ) : (
+          <Button disabled={true}>
+            <i className="prev" />
+            <span>Prev</span>
+          </Button>
+        )}
+        {next ? (
+          <LocalLink href={next} as={Button} disabled={!next}>
```
Comment (fabiosantoscode) [inline src/components/Documentation/NewMarkdown/index.js]: `next` here will always be truthy, so there should be no need to assign the disabled attribute.

File: src/html.js [hunk 6]
```diff
@@ -13,6 +13,14 @@ export default function HTML(props) {
...
           content="width=device-width, initial-scale=1, shrink-to-fit=no"
         />
+        <link
+          rel="stylesheet"
+          href="https://cdn.jsdelivr.net/npm/docsearch.js@2.6.2/dist/cdn/docsearch.min.css"
+        />
+        <script
+          type="text/javascript"
+          src="https://cdn.jsdelivr.net/npm/docsearch.js@2.6.2/dist/cdn/docsearch.min.js"
+        />
```
Comment (fabiosantoscode) [inline src/html.js]: Should we not import these files from our JS entry point to make them available in a single request?

**Comment by fabiosantoscode** [comment]
This is a gargantuan amount of work, good luck @iAdramelk :)

File: src/components/LocalLink/index.js (hunk 1 above)
Comment (iAdramelk) [reply src/components/LocalLink/index.js]: Good catch! I missed this part on the docs somehow. Will change it.

File: src/components/TrySection/index.js (hunk 2 above)
Comment (iAdramelk) [reply src/components/TrySection/index.js]: I probably should rename this component, all `Button`s on this page are actually `styled.a`, but yes, the all are rendered to `a`.

File: src/html.js (hunk 3 above)
Comment (iAdramelk) [reply src/html.js]: I'd like too, but we need to test it, to find out if it injects and executes code correctly, and that there is no problems if the code executes before html-element for it is rendered.

For now I'd like to leave it as it was in nextjs version, but maybe we should create separate task for that.

File: src/utils/ga.js (hunk 4 above)
Comment (iAdramelk) [reply src/utils/ga.js]: This, or maybe switch to GTM and send it from there.

File: src/components/Documentation/NewMarkdown/index.js (hunk 5 above)
Comment (iAdramelk) [reply src/components/Documentation/NewMarkdown/index.js]: You are right, will try to remove and check if there is some PropTypes errors. (I have vague memory that I left it there because of them, but I need to recheck)

File: src/html.js (hunk 6 above)
Comment (iAdramelk) [reply src/html.js]: I did it like that at first, but I started getting build errors with `window is undefined`, so for now I moved it back there. I'd like to research it a little later, probably in separate PR.

But I think that I can return css without problem.

Comment (fabiosantoscode) [reply src/html.js]: I see! This is probably due to serverside rendering.

In that case, there's 2 options:

1. Use the client entrypoint (maybe import it from gatsby-browser.js if you can't find it)
2. Use a conditional require. `if (typeof window === 'object') require('...')`.

Comment (fabiosantoscode) [reply src/html.js]: Either way, sure, it's not a big deal.

Comment (iAdramelk) [reply src/html.js]: Yep, it's definitely SSR. I thought about using `gatsby-browser.js` to fix it, but decided to leave it as is for now, until I make other stuff work.

**Comment by iAdramelk** [comment]
@fabiosantoscode I added AIP and reedirects middlewares + simple express server, can you please review them and advice on additional headers and other stuff that I maybe should add here?

**Comment by shcheklein** [comment]
@iAdramelk please rebase/merge, and let's check why CI is failing

File: src/components/Community/Section/index.js [hunk 7]
```diff
@@ -47,12 +45,6 @@ export default function CommunitySection({
     }
 
     updateVisibility()
-
-    Router.events.on('hashChangeComplete', updateVisibility)
-
-    return () => {
-      Router.events.off('hashChangeComplete', updateVisibility)
-    }
   }, [])
```
Comment (fabiosantoscode) [inline src/components/Community/Section/index.js]: This needs `anchor` as a dependency so it updates after the first render.

File: src/components/Documentation/Markdown/index.js [hunk 8]
```diff
@@ -1,279 +1,145 @@
...
-  return typeof title === 'string' && SLUG_REGEXP.test(title)
-}
-
-export function extractSlugFromTitle(title) {
-  // extracts expressions like {#too-many-files} from the end of a title
-  const meta = title.match(SLUG_REGEXP)
-
-  if (meta) {
-    return [title.substring(0, meta.index), meta[1]]
+const isInsideCodeBlock = elem => {
```
Comment (fabiosantoscode) [inline src/components/Documentation/Markdown/index.js]: You might want to have a look at [closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest).

File: src/components/Documentation/index.js [hunk 9]
```diff
@@ -1,50 +1,25 @@
...
-export default function Documentation({
-  source,
-  path,
-  next,
-  prev,
-  tutorials,
-  markdown
-}) {
-  const headings = useMemo(() => parseHeadings(markdown))
+  // const headings = useMemo(() => parseHeadings(markdown))
```
Comment (fabiosantoscode) [inline src/components/Documentation/index.js]: Forgotten comment?

File: src/html.js (hunk 6 above)
Comment (fabiosantoscode) [reply src/html.js]: Maybe in another PR. No big deal I think.

File: src/components/Documentation/NewMarkdown/index.js (hunk 5 above)
Comment (iAdramelk) [reply src/components/Documentation/NewMarkdown/index.js]: Removed it

File: server.js [hunk 10]
```diff
@@ -1,66 +1,16 @@
...
-        res.statusCode = 404
-      }
+app.use(redirectsMiddleware)
+app.use('/api', apiMiddleware)
 
-      // Custom route for all docs
-      app.render(req, res, '/doc', query)
-    } else {
-      // Regular Next.js handler
+app.use(express.static('public'))
```
[the remark this answers is not in the document]
Comment (iAdramelk) [reply server.js]: Added, thanks

File: gatsby-node.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      matches.push({
+        text: title,
+        slug: slug
+      })
+    }
+  } while (match)
+
+  slugger.reset()
+  return matches
+}
```
[the remark this answers is not in the document]
Comment (iAdramelk) [reply gatsby-node.js]: Yep, I just copied current implementation for the most part now. We should rework it in separate PR.

File: jest.config.js [hunk 12]
```diff
@@ -5,5 +5,6 @@ module.exports = {
   testEnvironment: 'node',
   transform: {
     '^.+\\.js?$': 'babel-jest'
-  }
+  },
+  testPathIgnorePatterns: ['/node_modules/', '/.cache']
```
[the remark this answers is not in the document]
Comment (iAdramelk) [reply jest.config.js]: Good catch, added it to ignore list.

File: middlewares/api/index.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+/* eslint-env node */
```
[the remark this answers is not in the document]
Comment (iAdramelk) [reply middlewares/api/index.js]: Fixed, thanks!

File: middlewares/redirects/index.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  if (redirectLocation) {
+    // HTTP redirects
+
+    const queryStr = stringify(query)
+    if (queryStr) {
+      redirectLocation += '?' + queryStr
+    }
+    res.writeHead(redirectCode, {
+      Location: redirectLocation
+    })
```
[the remark this answers is not in the document]
Comment (iAdramelk) [reply middlewares/redirects/index.js]: We should add `res.end()` here, but we still need to run `next()` after that block for the cases that are not redirected and should be handled by static middleware.

[the remark this answers is not in the document]
Comment (fabiosantoscode) [reply middlewares/redirects/index.js]: You still need to return early here though :snowman: . Otherwise, the next middleware in line will run, which we don't want since `res.end()` has already been called.

[the remark this answers is not in the document]
Comment (iAdramelk) [reply middlewares/redirects/index.js]: Yes, you are right, I forgot about it.

File: .gitignore [hunk 15]
```diff
@@ -56,8 +56,11 @@ typings/
...
 .idea
-.next/
 *.log
 
 # Mac finder artifacts
 .DS_Store
+
+# Gatsby cache
+.cache
+public
```
Comment (shcheklein) [inline .gitignore]: minor: newline

File: config/prismjs/dvc-commands.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  'remote add',
+  'remote',
+  'push',
+  'pull',
+  'pkg',
+  'pipeline show',
+  'pipeline list',
+  'pipeline',
+  'move',
+  'metrics show',
```
Comment (shcheklein) [inline config/prismjs/dvc-commands.js]: + metrics diff

File: config/prismjs/dvc-commands.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  'metrics remove',
+  'metrics modify',
+  'metrics add',
+  'metrics',
+  'lock',
+  'install',
+  'init',
+  'import-url',
```
Comment (shcheklein) [inline config/prismjs/dvc-commands.js]: + list

File: README.md [hunk 18]
```diff
@@ -1,4 +1,4 @@
-![owl logo](https://dvc.org/static/img/logo-github-readme.png)
+![owl logo](https://dvc.org/img/logo-github-readme.png)
```
Comment (shcheklein) [inline README.md]: why removing static prefix?

File: content/docs/install/completion.md [hunk 19]
```diff
@@ -134,7 +134,7 @@ $ exec $SHELL -l
 This step is optional but will make the DVC output look much nicer, by adding
 more colors to it. Add the following to your `~/.zshrc`:
 
-```zsh
+```bash
```
Comment (shcheklein) [inline content/docs/install/completion.md]: why this change? it is zsh, right?

Comment (iAdramelk) [reply content/docs/install/completion.md]: Basically because both `hljs` and `prsim` actually didn't have syntaxes for `zsh` or `sh` and they fallback for default mode then you use it.

See: https://github.com/conorhastings/react-syntax-highlighter/tree/master/src/languages/hljs
And https://github.com/conorhastings/react-syntax-highlighter/tree/master/src/languages/prism

And we also are not importing them it https://github.com/iterative/dvc.org/blob/master/src/components/Documentation/Markdown/index.js either.

Default mode for `hljs` here is something c-like and is highlight some things as a result. But `prismjs` that `gatsby` uses just leave code unhighlighted.  So here I have two options – either write `zsh` syntax from scratch (but we only use it in this one place) or try to find something close enough to the default `hljs` syntax. I choose later for now.

**Comment by jorgeorpinel** [comment]
Hello! Been trying to catch up with the review here. I have one intro Q. When I click on BLOG in the review app (https://dvc-landing-gatsby-migr-abefpq.herokuapp.com/features) it goes straight to the blog on production. Shouldn't it render the blog in the review app so we can check it since it's all integrated now? Thanks

File: content/docs/install/completion.md (hunk 19 above)
Comment (shcheklein) [reply content/docs/install/completion.md]: sounds good, @iAdramelk ! thanks.

File: gatsby-config.js [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+          options: {
+            enableCustomId: true,
+            isIconAfterHeader: true
+          }
+        }
+      ]
+    }
+  },
+  'gatsby-plugin-catch-links',
+  /*{
```
Comment (shcheklein) [inline gatsby-config.js]: remove commented code

**Comment by jorgeorpinel** [comment]
> edit on Github still does not work (from https://github.com/iterative/dvc.org/pull/1025#pullrequestreview-370574242)

It goes to master branch so that will be fixed when this is merged, right?

- The DOC link seems to be hard coded to /doc/get-started ? It's /doc currently
- /doc path should open the first doc (currently the Get-Started) without redirecting. It returns a 404 now. This is important for redirects (not just for the DOC link above).

**Redirects**
- How can we test the subdomain redirects? (www, man, error, code/data/remote)
  > Also, we should consider that redirects.test.js is always testing dvc.org, not the review app.
- URL path redirects e.g. /documentation/command-reference add a trailing slash e.g. -> /doc/command-reference/, which causes a 2nd redirect.

> NEXT on https://dvc-landing-gatsby-migr-abefpq.herokuapp.com/doc/command-reference/version does not work

This is still happening. Refers to the image button 🙂

File: gatsby-node.js [hunk 21]
```diff
@@ -0,0 +1,106 @@
+/* eslint-env node */
+
+const path = require('path')
+const GithubSlugger = require('github-slugger')
+
+const { getItemBySource } = require('./src/utils/sidebar')
+
+const slugger = new GithubSlugger()
+
+// Generate hedinga data from markdown
```
Comment (shcheklein) [inline gatsby-node.js]: typo: headings

File: package.json [hunk 22]
```diff
@@ -43,22 +47,25 @@
...
     "react-collapse": "^5.0.1",
     "react-collapsible": "^2.6.2",
     "react-dom": "^16.12.0",
     "react-ga": "^2.7.0",
+    "react-helmet": "^5.2.1",
     "react-markdown": "^4.2.2",
     "react-popover": "^0.5.10",
     "react-scroll": "^1.7.13",
     "react-slick": "^0.25.2",
     "react-syntax-highlighter": "^11.0.2",
```
Comment (shcheklein) [inline package.json]: do we need this? is there a way to cleanup the dependencies?

Comment (iAdramelk) [reply package.json]: We can safely remove `react-syntax-highlighter` here, yes, as well as `react-markdown` and some other libraries.

File: src/components/Documentation/Markdown/index.js [hunk 23]
```diff
@@ -1,279 +1,145 @@
...
+    <Content id="markdown-root">
+      {tutorials && (
+        <TutorialsWrapper>
+          <Tutorials tutorials={tutorials} compact={true} />
+        </TutorialsWrapper>
+      )}
+      <GithubLink href={githubLink} target="_blank" rel="noreferrer noopener">
+        <i /> Edit on GitHub
+      </GithubLink>
+      <div className="markdown-body">{renderAst(htmlAst)}</div>
```
Comment (shcheklein) [inline src/components/Documentation/Markdown/index.js]: so, what is the plugin that holding us from the danger html?

Comment (iAdramelk) [reply src/components/Documentation/Markdown/index.js]: Basically those two blocks: https://github.com/iterative/dvc.org/blob/5f589deee8fa32aca0f8c498947649fa02a87099/src/components/Documentation/Markdown/index.js#L31-L57

Comment (shcheklein) [reply src/components/Documentation/Markdown/index.js]: it should be straightforward similar to what we did already on the blog side a few times?

File: src/components/Documentation/Markdown/styles.js [hunk 24]
```diff
@@ -31,6 +31,131 @@ export const Content = styled.article`
...
+      position: relative;
+      top: 1px;
+      right: 0;
+      width: 12px;
+      height: 12px;
+      margin-left: 1px;
+      content: url(/img/external-link.svg);
+    }
+
+    pre[class*='language-'] {
```
Comment (shcheklein) [inline src/components/Documentation/Markdown/styles.js]: we will need to share it anyway? move it to a separate file?

File: src/components/Documentation/Markdown/index.js (hunk 23 above)
Comment (iAdramelk) [reply src/components/Documentation/Markdown/index.js]: It's a little harder here. We just replaced one html with another in the blog. But for example `<Tooltip>` is a custom react component with a lot of custom logic for positioning. And `<Collapsible>` is using amitations and react state to show and hide content. We can't just replace them with static html.

Comment (shcheklein) [reply src/components/Documentation/Markdown/index.js]: that's sad ... mostly means JS bundle will be way bigger? did you compare it already? Also means probably that we spent quite a lot of time on the blog side for no reason - they will be using the same JS, right?

File: src/components/Documentation/SidebarMenu/index.js [hunk 25]
```diff
@@ -5,24 +5,37 @@ import scrollIntoView from 'dom-scroll-into-view'
...
+  source,
+  path,
+  activePaths,
+  onClick
+}) {
   const isActive = activePaths && includes(activePaths, path)
   const isRootParent =
     activePaths && activePaths.length > 1 && activePaths[0] === path
 
+  // ToDo: Dirty hack to make links with source=false work
```
Comment (shcheklein) [inline src/components/Documentation/SidebarMenu/index.js]: create a ticket or put it to some other visible place

File: src/components/Documentation/index.js [hunk 26]
```diff
@@ -1,50 +1,24 @@
 /* global docsearch:readonly */
```
Comment (shcheklein) [inline src/components/Documentation/index.js]: why do we have it in some files?

File: src/components/Footer/index.js [hunk 27]
```diff
@@ -16,6 +18,8 @@ import {
   Wrapper
 } from './styles'
 
+const docsPage = getFirstPage()
```
Comment (shcheklein) [inline src/components/Footer/index.js]: means that `https://dvc-landing-gatsby-migr-abefpq.herokuapp.com/doc/` does not work

File: src/html.js (hunk 6 above)
Comment (shcheklein) [reply src/html.js]: please, create a ticket for this

File: README.md (hunk 18 above)
Comment (iAdramelk) [reply README.md]: Gatsby use /public/static/` folder to store autogenerated images and other assets. I decided that using the same name can cause possible new bugs with name confilcts and can make checking the build results harder, so I removed `/static/` part.

Comment (iAdramelk) [reply README.md]: Oops, my error. Forgot that `gatsby` actually automatically remove it. We still have folder `/static/img` in repo, but `gatsby` copies it's content to the root of the `/public/` folder. I could add nested folder `/static/static/img`, but I decide that it will be strange.

File: src/components/Documentation/Markdown/styles.js (hunk 24 above)
Comment (iAdramelk) [reply src/components/Documentation/Markdown/styles.js]: We will. but we will first convert `styled-components` to something else. If we move it right now we will still need to rewrite it later so I leave it this way for now.

File: src/components/Documentation/index.js (hunk 26 above)
Comment (iAdramelk) [reply src/components/Documentation/index.js]: In this case, because we connect `docsearch` here https://github.com/iterative/dvc.org/pull/1025/files#diff-58826c9c508b82875d8f9a9872d19e2dR18 and it register global variable `window.docsearch`. Eslint didn't know about this global variable so we tell him about it here.

File: src/components/Footer/index.js (hunk 27 above)
Comment (iAdramelk) [reply src/components/Footer/index.js]: We need to add it to the redirects list, yes. Having two identical pages on different urls probably not good for SEO too.

Comment (iAdramelk) [reply src/components/Footer/index.js]: Added redirects for all pages with `source: false`.

File: src/components/Documentation/SidebarMenu/index.js (hunk 25 above)
Comment (iAdramelk) [reply src/components/Documentation/SidebarMenu/index.js]: Done https://github.com/iterative/dvc.org/blob/5f589deee8fa32aca0f8c498947649fa02a87099/src/components/Documentation/SidebarMenu/index.js#L31

File: src/components/Documentation/Markdown/index.js (hunk 23 above)
Comment (iAdramelk) [reply src/components/Documentation/Markdown/index.js]: Not exaclty. We remove around 90% of code and libraries that we have before here: `react-markdown` (21kb minified), `react-syntax-hightlighter` (hard to say exactly, all themes and code is 397kb minified, but we probably used only 50-150kb in our case), `unist-util-visit` and our custom plugins for links and custom syntaxes. It's a lot of code.

We added new library `rehype-react` that have weight of `7.7kb` gzipped which is not perfect, but bearable and increased size of page data json because we are now sending not html, but parsed html ast. But these changes again are partially mitigated by gzip because ast is easy to compress. So in the end we have more code than we could if we just use static html, but significantly less then before migration.

Also blog pages will not be affected by it. Gatsby and next both create separate js-chunks for different page types, so `rehype-react` will only be inside doc page chunk.

File: package.json (hunk 22 above)
Comment (iAdramelk) [reply package.json]: Ok. Removed `react-syntax-highlighter` but didn't remove `react-markdown`. I forgot that it is used in Tooltip. Created ticket for removal https://github.com/iterative/dvc.org/issues/1045

File: gatsby-node.js (hunk 21 above)
Comment (iAdramelk) [reply gatsby-node.js]: Fixed

**Comment by iAdramelk** [comment]
> Hello! Been trying to catch up with the review here. I have one intro Q: When I click on BLOG in the review app (https://dvc-landing-gatsby-migr-abefpq.herokuapp.com/features) it goes straight to the blog on production. Shouldn't it render the blog in the review app so we can check it since it's all integrated now? Thanks

We only converted site itself right now. Blog is still in separate repo. :(



> edit on Github still does not work
> It goes to master branch so that will be fixed when this is merged, right?

Yes, it should work after we merge this PR to master.

File: gatsby-config.js (hunk 20 above)
Comment (iAdramelk) [reply gatsby-config.js]: Uncommented it actually, and added favicon, totally forgot about it.

File: config/prismjs/dvc-commands.js (hunk 17 above)
Comment (iAdramelk) [reply config/prismjs/dvc-commands.js]: Done

File: config/prismjs/dvc-commands.js (hunk 16 above)
Comment (iAdramelk) [reply config/prismjs/dvc-commands.js]: Done

File: .gitignore (hunk 15 above)
Comment (iAdramelk) [reply .gitignore]: Done

File: src/components/Documentation/index.js (hunk 9 above)
Comment (iAdramelk) [reply src/components/Documentation/index.js]: Yep, removed it

File: middlewares/redirects/index.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (iAdramelk) [reply middlewares/redirects/index.js]: Done

File: src/components/Documentation/Markdown/index.js (hunk 8 above)
Comment (iAdramelk) [reply src/components/Documentation/Markdown/index.js]: Just copied block for now too. Probably we need to update it separately.

File: src/html.js (hunk 6 above)
Comment (iAdramelk) [reply src/html.js]: Done https://github.com/iterative/dvc.org/issues/1046

File: README.md (hunk 18 above)
Comment (shcheklein) [reply README.md]: yeah, my concern is to keep as less URLs "locked" as possible. If we can get it into static it would be better (better w/o static/static of course). Also, I was thinking that we apply some image processor like in blog - to unify it, unify plugins, etc?

File: src/components/Documentation/Markdown/styles.js (hunk 24 above)
Comment (shcheklein) [reply src/components/Documentation/Markdown/styles.js]: kk, resolving 👍

File: README.md (hunk 18 above)
Comment (iAdramelk) [reply README.md]: Yep. Perfectly we should move images to corresponding components' folders and import them as other assets, this way will webpack automatically optimize images for us and rename them, move them correctly.

Comment (shcheklein) [reply README.md]: to corresponding "md"-files' folders you mean, right?

Comment (iAdramelk) [reply README.md]: Some to the md files, some to the components (like the icons in `/static/img/community/`).

Comment (shcheklein) [reply README.md]: 👍

File: package.json (hunk 22 above)
Comment (shcheklein) [reply package.json]: what about other libraries?

File: src/components/Documentation/Markdown/index.js (hunk 23 above)
Comment (shcheklein) [reply src/components/Documentation/Markdown/index.js]: yeah, I'm just wondering would have it bern then easier on the blog end to the same with plugins - we spent quite a lot of time writing them ... or am I missing something? at least I know this for the future decisions ...

File: package.json (hunk 22 above)
Comment (iAdramelk) [reply package.json]: `unist-util-visit` is still used, but now on server instead of client. Same with `isomorphic-fetch` (we still using it for API). Need to check others in more details, but at the first glance we still need them.

File: src/components/Footer/index.js (hunk 27 above)
Comment (shcheklein) [reply src/components/Footer/index.js]: > probably not good for SEO too.

yes, it's pretty bad

> We need to add it to the redirects list

make sure that it is 302 or 307 or whatever we use for temporary ones - we don't want them to be cached. we need to have a separate index page for dvc.org/doc

File: src/components/Documentation/index.js (hunk 26 above)
Comment (shcheklein) [reply src/components/Documentation/index.js]: 👍

File: src/components/Documentation/Markdown/index.js (hunk 23 above)
Comment (iAdramelk) [reply src/components/Documentation/Markdown/index.js]: Hard to say for sure. Simple tag replacement can probably be made simpler in browser but would increase bundle size. But stuff like that we did for image sizes are better implemented on the server.

File: src/components/Footer/index.js (hunk 27 above)
Comment (iAdramelk) [reply src/components/Footer/index.js]: Added 307 status code.

**Comment by fabiosantoscode** [comment]
I've noticed the preview deployment doesn't include the latest cache headers. Is this just a lack of rebasing, or was something lost in a merge?

The `cache-control` header needs a `s-maxage=999999` bit, so that the browser can validate against cloudflare, which is faster and doesn't hit our supple express server.

After each deployment, cloudflare is cleared (`npm start` runs `scripts/clear-cloudflare-cache.js`), so there should be no problem with this.

**Comment by fabiosantoscode** [comment]
I also noticed the 404 page is `text/plain`. Gatsby should output a `public/404.html` file, which can be used to handle 404's.

I think this can be done with an express error handler at the end of the chain. As far as I can see `express.static` doesn't provide an option for a 404 page.

**Comment by pavelgrinchenko** [comment]
> I've noticed the preview deployment doesn't include the latest cache headers. Is this just a lack of rebasing, or was something lost in a merge?

Added `Cache-Control` header

> I also noticed the 404 page is text/plain. Gatsby should output a public/404.html file, which can be used to handle 404's.

Added 404 page which is similar to the one in the blog

File: src/components/Nav/index.js [hunk 28]
```diff
@@ -4,10 +4,14 @@ import PropTypes from 'prop-types'
 import LocalLink from '../LocalLink'
 
 import { logEvent } from '../../utils/ga'
+import { getFirstPage } from '../../utils/sidebar'
+
+const docsPage = getFirstPage()
```
Comment (shcheklein) [inline src/components/Nav/index.js]: do we still need this? it should be always `dvc.org/doc`

File: src/pages/404.js [hunk 29]
```diff
@@ -0,0 +1,18 @@
+import React from 'react'
```
Comment (shcheklein) [inline src/pages/404.js]: do we need both 404 pages? I thought I saw the second one ..

Comment (shcheklein) [reply src/pages/404.js]: @pavelgrinchenko ? ^^

File: src/components/Nav/index.js (hunk 28 above)
Comment (shcheklein) [reply src/components/Nav/index.js]: @pavelgrinchenko ? ^^

File: README.md (hunk 18 above)
Comment (shcheklein) [reply README.md]: @iAdramelk please, create a ticket for this as well

File: src/pages/404.js (hunk 29 above)
Comment (pavelgrinchenko) [reply src/pages/404.js]: Do you mean `src/pages/404.js` and `src/components/404/`?
In first case it file for Gatsby which tells to it to generate page `404.html`. In the second case it's React component which will be rendered on this page, with all styles that it has. Other pages in the `src/pages' folder have the same structure.

File: src/components/Nav/index.js (hunk 28 above)
Comment (pavelgrinchenko) [reply src/components/Nav/index.js]: I tried to remove this helper and rely on server.js redirects, but they work just for the initial load. For example, we load `/doc` page in the browser - in this case, server.js redirects us to the correct page. But if we are navigating from another page of dvc.org then Gatsby's router will intercept this request and show the 404 page, because in this case, it doesn't know about any redirects on the server side.
@iAdramelk told me that you are working on an index page for `/doc`. So we can wait for it and then remove this helper.
Another solution is to use something like [this plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-client-side-redirect/), which will create all redirects on the client-side too, to have identical behavior on both sides: client and server.

Comment (shcheklein) [reply src/components/Nav/index.js]: @pavelgrinchenko makes sense. it might be that we have a problem for other similar pages (we have 3 or 4 pages like this) if we have links to them in the docs?

> Another solution is to use something like this plugin, which will create all redirects on the client-side too, to have identical behavior on both sides: client and server.

I don't like the, but they might be good as a fallback for those sections we won't have index pages for some reason.

File: src/components/DocLayout/SidebarMenu/index.js [hunk 30]
```diff
@@ -76,14 +76,18 @@ export default function SidebarMenu({ id, sidebar, currentPath, onClick }) {
       })
     }
 
-    const node = document.getElementById(currentPath)
+    const node = document.getElementById(currentPath.replace(/\/$/, ''))
```
Comment (fabiosantoscode) [inline src/components/DocLayout/SidebarMenu/index.js]: This should be handled at the serverside I believe. It should act as the gatekeeper which prevents us from having to deal with the trailing-slash-case on the client.

File: src/pages/404.js (hunk 29 above)
Comment (shcheklein) [reply src/pages/404.js]: 👍

File: src/components/DocLayout/SidebarMenu/index.js (hunk 30 above)
Comment (fabiosantoscode) [reply src/components/DocLayout/SidebarMenu/index.js]: We can add a redirect to redirects-list.json which redirects from `^(.+)/$` to `$1` and that should do it. We never want trailing slashes.

@iAdramelk told me this breaks the static server even if its [redirects](https://github.com/expressjs/serve-static#redirect) option is disabled, but we can always use [serve-handler](https://npmjs.com/serve-handler). The choice doesn't matter much. This is all going in S3 eventually.

File: README.md (hunk 18 above)
Comment (iAdramelk) [reply README.md]: Done https://github.com/iterative/dvc.org/issues/1063

Comment (shcheklein) [reply README.md]: thanks!

**Comment by shcheklein** [comment]
> /doc path redirects to /doc/get-started ... The DOC link still seems to be hardcoded to /doc/get-started

@jorgeorpinel there is a ticket to handle all of this after the merge - https://github.com/iterative/dvc.org/issues/1060 - please chime in here. 

> How can we test the subdomain redirects? (www, man, error, code/data/remote)

we will deploy and test them

> which causes a 2nd redirect to the final location: /doc/command-reference/

will be handled when we merge dvc.org with blog the way I understand this cc @fabiosantoscode @iAdramelk @pavelgrinchenko

File: src/components/DocLayout/SidebarMenu/index.js (hunk 30 above)
Comment (pavelgrinchenko) [reply src/components/DocLayout/SidebarMenu/index.js]: I'll take a look a `serve-handler` tomorrow.
But we also discussed this issue [yesterday at Slack](https://evilmartians.slack.com/archives/CLN9J8Q8K/p1584377665016000?thread_ts=1584348448.007700&cid=CLN9J8Q8K) and decided that it can be resolved after merging with blog's codebase.

**Comment by iAdramelk** [comment]
@jorgeorpinel 

> /doc path redirects to /doc/get-started. Again, what if we want to change the first doc to something else? Would we have to change the redirects file every time?

It's actually not hardcoded now. It resolves to the first item with the source in the `sidebar.json`.

**Comment by jorgeorpinel** [comment]
> It's actually not hardcoded now

I'm seeing it hardcoded in redirects-list.json @iAdramelk

**Comment by iAdramelk** [comment]
> I'm seeing it hardcoded in redirects-list.json

It's hardcoded in redirects for now (to redirect users who hit urls from search engines or bookmarks), but actual links in html and site header/footer are not hardcoded. They are rosolved using sidebar.json. 

But you are right, it's not an ideal situation and it can break if we change root sidebar.json right now without changing redirects.
